### PR TITLE
cls/rgw: fix the bug when trim usage by specifying user

### PR DESCRIPTION
**Pull request by lcwlovefamily**
when trim usage by specifying user, it is possible to trim other users’ usage.

Fixes:https://tracker.ceph.com/issues/48208

Signed-off-by: chengwu liang liangchengw@chinatelecom.cn
New PR from https://github.com/ceph/ceph/pull/38035

### CODE REVIEW

**Comment by lcwlovefamily** [comment]
@batrick @tchaikov @jdurgin please take a look,thanks

**Comment by lcwlovefamily** [comment]
@batrick @tchaikov @jdurgin

File: src/cls/rgw/cls_rgw.cc [hunk 1]
```diff
@@ -3170,17 +3170,19 @@ static int usage_iterate_range(cls_method_context_t hctx, uint64_t start, uint64
...
     }
 
-    if (by_user && key.compare(0, user_key.size(), user_key) != 0) {
+    ret = usage_record_decode(iter->second, e);
+    if (ret < 0) {
+        return ret;
+    }
+        
+    rgw_user* puser = (e.payer.empty() ? &e.owner : &e.payer);  // copy from rgw_user_usage_log_add
+    if (by_user && user_key != puser->to_str()) {
```
Comment (mattbenjamin) [inline src/cls/rgw/cls_rgw.cc]: isn't this stopping the iteration--and returning a value for the current iterator position--when the users do not match?

File: src/cls/rgw/cls_rgw.cc [hunk 2]
```diff
@@ -3170,13 +3170,13 @@ static int usage_iterate_range(cls_method_context_t hctx, uint64_t start, uint64
       return 0;
     }
 
-    ret = usage_record_decode(iter->second, e);
+    ret =  usage_record_decode(iter->second, e);
```
Comment (mattbenjamin) [inline src/cls/rgw/cls_rgw.cc:3173]: this ws change is unecessary

File: src/cls/rgw/cls_rgw.cc (hunk 1 above)
Comment (lcwlovefamily) [reply src/cls/rgw/cls_rgw.cc]: wouldn't happen. when i search testuser's usage i will have plenty omap keys like testuser_0XXX, testuser__0XXX,becauser digital is less than letter '_',so the correct user's usage is always in the first half；when i search testuser_'s usage i only get omap keys like testuser__0XXX without omap key like testuser_0XXX.i didnt change the return logic here which assume that the correct answer is always in the first half.

File: src/cls/rgw/cls_rgw.cc (hunk 2 above)
Comment (lcwlovefamily) [reply src/cls/rgw/cls_rgw.cc:3173]: ignore this, forgot to delete it

**Pull request closed without merging**